src/server/pps/server/api_server.go

```diff
@@ -2362,27 +2369,96 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
+			// since the spec commit was created
+			if err := a.sudo(txnCtx.Client, func(superClient *client.APIClient) error {
+				commitInfo, err := superClient.InspectCommit(ppsconsts.SpecRepo, pipelineName)
+				if err != nil {
+					return err
+				}
+				for _, child := range commitInfo.ChildCommits {
+					if child.ID == specCommit.ID {
+						return nil
+					}
+				}
+
+				// something has changed, try to delete this commit and make a new one
+				superClient.DeleteCommit(ppsconsts.SpecRepo, specCommit.ID)
+				specCommit = nil
+				if prevSpecCommit != nil {
+					*prevSpecCommit = nil
+				}
+				return nil
+			}); err != nil {
+				return specCommit, err
+			}
+		}
+
+		if specCommit == nil {
+			// make a commit branching off of master on a new, throwaway branch
+			tempBranch := testutil.UniqueString(pipelineName + "_")
+			if err := a.sudo(txnCtx.Client, func(superClient *client.APIClient) error {
+				// strip transaction so that commits happen for real
+				superClient = superClient.WithoutTransaction()
+				var err error
+				if update {
+					_, err = superClient.StartCommitParent(ppsconsts.SpecRepo, tempBranch, pipelineName)
+				} else {
+					_, err = superClient.StartCommit(ppsconsts.SpecRepo, tempBranch)
+				}
+				if err != nil {
+					return err
+				}
+				specCommit, err = a.makePipelineInfoCommitOnBranch(superClient, pipelineInfo, tempBranch)
+				if err != nil {
+					return err
+				}
+				err = superClient.FinishCommit(ppsconsts.SpecRepo, specCommit.ID)
+				if err != nil {
+					specCommit = nil // don't use the open commit
+					return err
+				}
+				// don't bother failing on error here, we won't be able to clean it up later, either
+				_ = superClient.DeleteBranch(ppsconsts.SpecRepo, tempBranch, false)
+				return nil
+			}); err != nil {
+				return nil, err
+			}
+		}
+
+		if prevSpecCommit != nil {
+			// overwrite with current commit pointer
+			*prevSpecCommit = specCommit
+		}
+		return specCommit, nil
 	}
 
 	if update {
 		// Help user fix inconsistency if previous UpdatePipeline call failed
-		if ci, err := pachClient.InspectCommit(ppsconsts.SpecRepo, pipelineName); err != nil {
-			return nil, err
+		if ci, err := txnCtx.Client.InspectCommit(ppsconsts.SpecRepo, pipelineName); err != nil {
+			return err
 		} else if ci.Finished == nil {
-			return nil, errors.Errorf("the HEAD commit of this pipeline's spec branch " +
+			return errors.Errorf("the HEAD commit of this pipeline's spec branch " +
 				"is open. Either another CreatePipeline call is running or a previous " +
 				"call crashed. If you're sure no other CreatePipeline commands are " +
 				"running, you can run 'pachctl update pipeline --clean' which will " +
 				"delete this open commit")
 		}
 
-		// Remove provenance from existing output branch, so that creating a new
-		// spec commit doesn't create an output commit in the old output branch.
-		if err := a.hardStopPipeline(pachClient, pipelineInfo); err != nil {
```

Comment:
I still need to work out the right way to close open commits. Straightforwardly converting the logic here into transaction versions still seems like it could miss jobs that aren't created yet, unless I'm missing some defense against that.